ufl/algorithms/comparison_checker.py

```diff
@@ -0,0 +1,150 @@
+            raise ComplexComparisonError("You can't compare complex numbers with gt.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def lt(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with lt.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def ge(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with ge.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def le(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with le.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def max_value(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with max.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def min_value(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with min.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def real(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'real'
+        return o
+
+    def imag(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'real'
+        return o
+
+    def sqrt(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'complex'
+        return o
+
+    def power(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        if float(ops[1]) < 1.0 and float(ops[1]) > 0.0:
+            self.nodetype[o] = 'complex'
+        elif self.nodetype[ops[0]] == 'complex':
+            self.nodetype[o] = 'complex'
+        else:
+            self.nodetype[o] = 'real'
+        return o
+
+    def abs(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'real'
+        return o
+
+    def terminal(self, term, *ops):
+        # default terminals to complex, except the ones we *know* are real
+        if type(term) in {IntValue, FloatValue, Zero, Argument}:
```

Comment:
Ugh, `isinstance` should be preferred unless there is a serious reason.

Reply:
Why is `Argument` here but `Coefficient` not? Ah, because we know that all basis functions are real valued...